stock_picking_group_by_partner_by_carrier/report/report_delivery_slip.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2021 Camptocamp
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)
+
+from odoo import api, models
+from odoo.tools import float_compare, float_round
+
+
+class DeliverySlipReport(models.AbstractModel):
+    _name = "report.stock.report_deliveryslip"
+    _description = "Delivery Slip Report"
+
+    @api.model
+    def get_remaining_to_deliver_data(self, moves):
+        """Return dictionaries encoding pending quantities to deliver
+
+        Returns a list of dictionaries, encoding the data to be displayed
+        at the end of the delivery slip, summarising for each order the
+        pending quantities for each of its lines.
+        """
+        remaining_to_deliver_data = []
+        for move in moves:
+            if not move.id:  # Header moves are created as mock moves.
+                remaining_to_deliver_data.append(
+                    {"is_header": True, "concept": move.description_picking}
+                )
+            elif (
+                move.picking_id.picking_type_id.code == "outgoing" and move.sale_line_id
+            ):
+                sale_order_line = move.sale_line_id
+                qty_expected = sale_order_line.product_uom._compute_quantity(
+                    sale_order_line.product_uom_qty, move.product_uom
+                )
+                qty_delivered = sale_order_line.product_uom._compute_quantity(
+                    sale_order_line.qty_delivered, move.product_uom
+                )
+                if (
+                    float_compare(
+                        qty_expected,
+                        qty_delivered,
+                        precision_rounding=move.product_uom.rounding,
+                    )
+                    != 0
+                ):
```

Comment:
maybe qty_to_deliver field already computes this data

Reply:
This computation was wrong, the code has been redone.